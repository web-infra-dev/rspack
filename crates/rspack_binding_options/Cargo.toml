[package]
edition    = "2021"
license    = "MIT"
name       = "rspack_binding_options"
repository = "https://github.com/web-infra-dev/rspack"
version    = "0.1.0"

[features]
default = []
plugin  = ["rspack_loader_swc/plugin"]

[dependencies]
async-trait                           = { workspace = true }
better_scoped_tls                     = { workspace = true }
derivative                            = { workspace = true }
glob                                  = { workspace = true }
napi                                  = { workspace = true, features = ["async", "tokio_rt", "serde-json", "anyhow"] }
napi-derive                           = { workspace = true }
rspack_binding_macros                 = { path = "../rspack_binding_macros" }
rspack_binding_values                 = { path = "../rspack_binding_values" }
rspack_core                           = { path = "../rspack_core" }
rspack_error                          = { path = "../rspack_error" }
rspack_identifier                     = { path = "../rspack_identifier" }
rspack_ids                            = { path = "../rspack_ids" }
rspack_loader_react_refresh           = { path = "../rspack_loader_react_refresh" }
rspack_loader_runner                  = { path = "../rspack_loader_runner" }
rspack_loader_swc                     = { path = "../rspack_loader_swc" }
rspack_napi_shared                    = { path = "../rspack_napi_shared" }
rspack_plugin_asset                   = { path = "../rspack_plugin_asset" }
rspack_plugin_banner                  = { path = "../rspack_plugin_banner" }
rspack_plugin_copy                    = { path = "../rspack_plugin_copy" }
rspack_plugin_css                     = { path = "../rspack_plugin_css" }
rspack_plugin_devtool                 = { path = "../rspack_plugin_devtool" }
rspack_plugin_ensure_chunk_conditions = { path = "../rspack_plugin_ensure_chunk_conditions" }
rspack_plugin_entry                   = { path = "../rspack_plugin_entry" }
rspack_plugin_externals               = { path = "../rspack_plugin_externals" }
rspack_plugin_hmr                     = { path = "../rspack_plugin_hmr" }
rspack_plugin_html                    = { path = "../rspack_plugin_html" }
rspack_plugin_javascript              = { path = "../rspack_plugin_javascript" }
rspack_plugin_json                    = { path = "../rspack_plugin_json" }
rspack_plugin_library                 = { path = "../rspack_plugin_library" }
rspack_plugin_limit_chunk_count       = { path = "../rspack_plugin_limit_chunk_count" }
rspack_plugin_merge_duplicate_chunks  = { path = "../rspack_plugin_merge_duplicate_chunks" }
rspack_plugin_mf                      = { path = "../rspack_plugin_mf" }
rspack_plugin_progress                = { path = "../rspack_plugin_progress" }
rspack_plugin_real_content_hash       = { path = "../rspack_plugin_real_content_hash" }
rspack_plugin_remove_empty_chunks     = { path = "../rspack_plugin_remove_empty_chunks" }
rspack_plugin_runtime                 = { path = "../rspack_plugin_runtime" }
rspack_plugin_schemes                 = { path = "../rspack_plugin_schemes" }
rspack_plugin_split_chunks            = { path = "../rspack_plugin_split_chunks" }
rspack_plugin_swc_css_minimizer       = { path = "../rspack_plugin_swc_css_minimizer" }
rspack_plugin_swc_js_minimizer        = { path = "../rspack_plugin_swc_js_minimizer" }
rspack_plugin_warn_sensitive_module   = { path = "../rspack_plugin_warn_sensitive_module" }
rspack_plugin_wasm                    = { path = "../rspack_plugin_wasm" }
rspack_plugin_web_worker_template     = { path = "../rspack_plugin_web_worker_template" }
rspack_plugin_worker                  = { path = "../rspack_plugin_worker" }
rspack_regex                          = { path = "../rspack_regex" }
rspack_swc_visitors                   = { path = "../rspack_swc_visitors" }
rustc-hash                            = { workspace = true }
serde                                 = { workspace = true, features = ["derive"] }
serde_json                            = { workspace = true }
swc_config                            = { workspace = true }
swc_core                              = { workspace = true, default-features = false, features = ["ecma_transforms_react"] }
tokio                                 = { workspace = true, features = ["rt", "rt-multi-thread", "macros", "test-util", "parking_lot"] }
tracing                               = { workspace = true }
