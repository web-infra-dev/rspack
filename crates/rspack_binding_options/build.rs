use std::env;

fn main() {
  // Rebuild binding options if and only if it's built for crate `node_binding`
  if env::var("OUT_DIR")
    .expect("should exist")
    .contains("node_binding")
  {
    println!("cargo:rerun-if-changed=../node_binding");
  }
}
