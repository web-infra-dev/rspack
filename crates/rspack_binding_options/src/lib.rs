#![feature(try_blocks)]
#![feature(let_chains)]
mod options;
mod plugins;
pub use options::*;
