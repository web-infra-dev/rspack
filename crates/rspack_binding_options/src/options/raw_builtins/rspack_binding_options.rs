use std::sync::Arc;

use derivative::Derivative;
use napi::{bindgen_prelude::Either3, Either, Env, JsFunction};
use napi_derive::napi;
use rspack_napi_shared::{
  get_napi_env,
  threadsafe_function::{ThreadsafeFunction, ThreadsafeFunctionCallMode},
  JsRegExp, NapiResultExt,
};
use rspack_plugin_devtool::{
  Append, DevtoolPluginOptions, FallbackModuleFilenameTemplate, ModuleFilenameTemplate,
};
use serde::Deserialize;

type RawAppend = Either3<String, bool, JsFunction>;

#[inline]
fn default_append() -> Append {
  Append::Default
}

fn normalize_raw_append(raw: RawAppend) -> Append {
  match raw {
    Either3::A(str) => Append::String(str),
    Either3::B(_) => Append::Disabled,
    Either3::C(v) => {
      let fn_payload: napi::Result<ThreadsafeFunction<(), Option<String>>> = try {
        let env = get_napi_env();
        rspack_binding_macros::js_fn_into_threadsafe_fn!(v, &Env::from(env))
      };
      let fn_payload = fn_payload.expect("convert to threadsafe function failed");
      Append::Fn(Arc::new(move || {
        fn_payload
          .call((), ThreadsafeFunctionCallMode::NonBlocking)
          .into_rspack_result()
          .expect("into rspack result failed")
          .blocking_recv()
          .unwrap_or_else(|err| panic!("Failed to call external function: {err}"))
          .expect("failed")
      }))
    }
  }
}

type RawRule = Either<String, JsRegExp>;
type RawRules = Either<RawRule, Vec<RawRule>>;

type RawFallbackModuleFilenameTemplate = Either3<String, bool, JsFunction>;

#[inline]
fn default_fallback_module_filename_template() -> FallbackModuleFilenameTemplate {
  FallbackModuleFilenameTemplate::Disabled
}

fn normalize_raw_fallback_module_filename_template(
  raw: RawFallbackModuleFilenameTemplate,
) -> FallbackModuleFilenameTemplate {
  match raw {
    Either3::A(str) => FallbackModuleFilenameTemplate::String(str),
    Either3::B(_) => FallbackModuleFilenameTemplate::Disabled,
    Either3::C(v) => {
      let fn_payload: napi::Result<ThreadsafeFunction<(), Option<String>>> = try {
        let env = get_napi_env();
        rspack_binding_macros::js_fn_into_threadsafe_fn!(v, &Env::from(env))
      };
      let fn_payload = fn_payload.expect("convert to threadsafe function failed");
      FallbackModuleFilenameTemplate::Fn(Arc::new(move || {
        fn_payload
          .call((), ThreadsafeFunctionCallMode::NonBlocking)
          .into_rspack_result()
          .expect("into rspack result failed")
          .blocking_recv()
          .unwrap_or_else(|err| panic!("Failed to call external function: {err}"))
          .expect("failed")
      }))
    }
  }
}

type RawModuleFilenameTemplate = Either3<String, bool, JsFunction>;

#[inline]
fn default_module_filename_template() -> ModuleFilenameTemplate {
  ModuleFilenameTemplate::Disabled
}

fn normalize_raw_module_filename_template(
  raw: RawModuleFilenameTemplate,
) -> ModuleFilenameTemplate {
  match raw {
    Either3::A(str) => ModuleFilenameTemplate::String(str),
    Either3::B(_) => ModuleFilenameTemplate::Disabled,
    Either3::C(v) => {
      let fn_payload: napi::Result<ThreadsafeFunction<(), Option<String>>> = try {
        let env = get_napi_env();
        rspack_binding_macros::js_fn_into_threadsafe_fn!(v, &Env::from(env))
      };
      let fn_payload = fn_payload.expect("convert to threadsafe function failed");
      ModuleFilenameTemplate::Fn(Arc::new(move || {
        fn_payload
          .call((), ThreadsafeFunctionCallMode::NonBlocking)
          .into_rspack_result()
          .expect("into rspack result failed")
          .blocking_recv()
          .unwrap_or_else(|err| panic!("Failed to call external function: {err}"))
          .expect("failed")
      }))
    }
  }
}

#[derive(Derivative, Deserialize)]
#[serde(rename_all = "camelCase")]
#[napi(object)]
#[derivative(Debug)]
pub struct RawSourceMapDevToolPluginOptions {
  #[serde(skip_deserializing)]
  #[napi(ts_type = "(false | null) | string | Function")]
  #[derivative(Debug = "ignore")]
  pub append: Option<RawAppend>,
  pub columns: Option<bool>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | RegExp | (string | RegExp)[]")]
  #[derivative(Debug = "ignore")]
  pub exclude: Option<RawRules>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | false | Function")]
  #[derivative(Debug = "ignore")]
  pub fallback_module_filename_template: Option<RawFallbackModuleFilenameTemplate>,
  pub file_context: Option<String>,
  #[napi(ts_type = "(false | null) | string")]
  pub filename: Option<String>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | RegExp | (string | RegExp)[]")]
  #[derivative(Debug = "ignore")]
  pub include: Option<RawRules>,
  pub module: Option<bool>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | false | Function")]
  #[derivative(Debug = "ignore")]
  pub module_filename_template: Option<RawModuleFilenameTemplate>,
  pub namespace: Option<String>,
  pub no_sources: Option<bool>,
  pub public_path: Option<String>,
  pub source_root: Option<String>,
  #[serde(skip_deserializing)]
  #[napi(ts_type = "string | RegExp | (string | RegExp)[]")]
  #[derivative(Debug = "ignore")]
  pub test: Option<RawRules>,
}

impl From<RawSourceMapDevToolPluginOptions> for DevtoolPluginOptions {
  fn from(raw_opts: RawSourceMapDevToolPluginOptions) -> Self {
    let append = raw_opts
      .append
      .map_or(default_append(), |name| normalize_raw_append(name));

    DevtoolPluginOptions {
      append,
      inline: todo!(),
      namespace: todo!(),
      columns: todo!(),
      no_sources: todo!(),
      public_path: raw_opts.public_path,
    }
  }
}
