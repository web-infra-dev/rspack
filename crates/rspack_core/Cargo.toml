[package]
edition    = "2021"
license    = "MIT"
name       = "rspack_core"
repository = "https://github.com/web-infra-dev/rspack"
version    = "0.1.0"

[dependencies]
anymap = "1.0.0-beta.2"
async-recursion = { workspace = true }
async-trait = { workspace = true }
bitflags = { workspace = true }
dashmap = { workspace = true }
derivative = { workspace = true }
dyn-clone = "1.0.16"
either = "1"
futures = { workspace = true }
glob-match = "0.2.1"
hashlink = { workspace = true }
hex = "0.4.3"
indexmap = { workspace = true }
itertools = { workspace = true }
json = { workspace = true }
mime_guess = { workspace = true }
once_cell = { workspace = true }
oxc_resolver = { version = "1.5.3" }
paste = { workspace = true }
petgraph = { version = "0.6.4", features = ["serde-1"] }
rayon = { workspace = true }
regex = { workspace = true }
rkyv = { workspace = true, features = ["indexmap", "validation"] }
rspack_ast = { path = "../rspack_ast" }
rspack_core_macros = { path = "../rspack_core_macros" }
rspack_database = { path = "../rspack_database", features = ["rayon"] }
rspack_error = { path = "../rspack_error" }
rspack_fs = { path = "../rspack_fs", features = ["async", "rspack-error"] }
rspack_futures = { path = "../rspack_futures" }
rspack_hash = { path = "../rspack_hash" }
rspack_hook = { path = "../rspack_hook" }
rspack_identifier = { path = "../rspack_identifier" }
rspack_loader_runner = { path = "../rspack_loader_runner" }
rspack_regex = { path = "../rspack_regex" }
rspack_sources = { workspace = true }
rspack_swc_visitors = { path = "../rspack_swc_visitors" }
rspack_util = { path = "../rspack_util" }
rustc-hash = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
sugar_path = { workspace = true }
swc_core = { workspace = true, features = [
  "__ecma",
  "__visit",
  "__utils",
  "__ecma_transforms",
  "ecma_ast",
  "ecma_codegen",
  "ecma_quote",
  "common_concurrent",
  "common_tty",
  "swc_ecma_ast",
  "ecma_transforms_react",
  "ecma_transforms_module",
  "swc_ecma_codegen",
  "swc_ecma_visit",
  "ecma_visit_path",
  "css_ast",
  "css_codegen",
  "css_parser",
  "css_utils",
  "css_visit",
  "css_visit_path",
  "css_compat",
  "css_modules",
] }
swc_node_comments = { workspace = true }
tokio = { workspace = true, features = ["rt", "rt-multi-thread", "macros", "test-util", "parking_lot"] }
tracing = { workspace = true }
url = { workspace = true }
ustr = { workspace = true }

[dev-dependencies]
pretty_assertions = { version = "1.4.0" }
