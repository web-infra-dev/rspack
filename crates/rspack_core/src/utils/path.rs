use std::path::{Component, Path};
use sugar_path::{self, PathSugar};
pub fn normalize_path(path: &str, root: &str) -> String {
  let res = Path::new(&root)
    .relative(Path::new(&path))
    .to_string_lossy()
    .to_string();

  res
}

pub fn uri_to_chunk_name(root: &str, uri: &str) -> String {
  println!("root {:?}", root);
  let path = Path::new(uri);
  let mut relatived = Path::new(root).relative(path);
  let ext = relatived
    .extension()
    .and_then(|ext| ext.to_str())
    .unwrap_or("")
    .to_string();
  relatived.set_extension("");
  let mut name = relatived
    .components()
    .filter(|com| matches!(com, Component::Normal(_)))
    .filter_map(|seg| seg.as_os_str().to_str())
    .intersperse("_")
    .fold(String::new(), |mut acc, seg| {
      acc.push_str(seg);
      acc
    });
  name.push('_');
  name.push_str(&ext);
  name
}
