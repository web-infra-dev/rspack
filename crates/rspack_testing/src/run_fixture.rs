use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};

use cargo_rst::{helper::make_relative_from, rst::RstBuilder};
use insta::{assert_snapshot, Settings};
use itertools::Itertools;
use rspack_binding_options::{RawOptions, RawOptionsApply};
use rspack_core::{BoxPlugin, Compiler, CompilerOptions};
use rspack_fs::AsyncNativeFileSystem;
use rspack_tracing::enable_tracing_by_env;

use crate::{eval_raw::evaluate_to_json, test_config::TestConfig};

pub type MutTestOptionsFn = dyn FnMut(&mut Vec<BoxPlugin>, &mut CompilerOptions);

pub fn apply_from_fixture(fixture_path: &Path) -> (CompilerOptions, Vec<BoxPlugin>) {
  let js_config = fixture_path.join("test.config.js");
  if js_config.exists() {
    let raw = evaluate_to_json(&js_config);
    let raw: RawOptions = serde_json::from_slice(&raw).expect("ok");
    let mut plugins = Vec::new();
    let compiler_options = raw.apply(&mut plugins).expect("should be ok");
    return (compiler_options, plugins);
  }
  let json_config = fixture_path.join("test.config.json");
  let test_config = TestConfig::from_config_path(&json_config);
  test_config.apply(fixture_path.to_path_buf())
}
#[tokio::main]
pub async fn test_fixture_html(fixture_path: &Path) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".html"),
    Box::new(|_, _| {}),
    None,
  )
  .await
}
#[tokio::main]
pub async fn test_fixture_js(fixture_path: &Path) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".js") && !s.contains("runtime.js"),
    Box::new(|_, _| {}),
    None,
  )
  .await
}
#[tokio::main]
pub async fn test_fixture_css(fixture_path: &Path) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".css"),
    Box::new(|_, _| {}),
    None,
  )
  .await
}
#[tokio::main]
pub async fn test_fixture_css_modules(fixture_path: &Path) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".css") || (s.ends_with(".js") && !s.contains("runtime.js")),
    Box::new(|_, _| {}),
    None,
  )
  .await
}
#[tokio::main]
pub async fn test_fixture_insta(
  fixture_path: &Path,
  stats_filter: &dyn Fn(&str) -> bool,
  mut_settings: Box<MutTestOptionsFn>,
) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(fixture_path, stats_filter, mut_settings, None).await
}

#[tokio::main]
pub async fn test_fixture(
  fixture_path: &Path,
  mut_settings: Box<MutTestOptionsFn>,
  snapshot_name: Option<String>,
) -> Compiler<AsyncNativeFileSystem> {
  test_fixture_share(
    fixture_path,
    &|s| s.ends_with(".js") && !s.contains("runtime.js"),
    mut_settings,
    snapshot_name,
  )
  .await
}

/// You can mutate the `Settings` of insta and  `CompilerOptions` of rspack via `mut_settings` mut closure
pub async fn test_fixture_share(
  fixture_path: &Path,
  stats_filter: &dyn Fn(&str) -> bool,
  mut mut_settings: Box<MutTestOptionsFn>,
  snapshot_name: Option<String>,
) -> Compiler<AsyncNativeFileSystem> {
  enable_tracing_by_env(&std::env::var("TRACE").ok().unwrap_or_default(), "stdout");

  let snapshot_name = snapshot_name.unwrap_or("output".to_string());

  let mut settings = Settings::clone_current();
  settings.set_snapshot_path(Path::new(fixture_path).join("snapshot"));
  settings.set_omit_expression(true);
  settings.set_prepend_module_to_snapshot(false);

  let (mut options, mut plugins) = apply_from_fixture(fixture_path);

  mut_settings(&mut plugins, &mut options);

  // clean output
  if options.output.path.exists() {
    std::fs::remove_dir_all(&options.output.path).expect("should remove output");
  }
  let mut compiler = Compiler::new(options, plugins, AsyncNativeFileSystem);

  compiler
    .build()
    .await
    .unwrap_or_else(|e| panic!("failed to compile in fixture {fixture_path:?}, {e:#?}"));
  let assets = compiler.compilation.assets();

  let content = assets
    .iter()
    .filter_map(|(filename, asset)| {
      if stats_filter(filename) {
        println!("{:?}", settings.description());
        let content = asset
          .get_source()
          .map(|x| x.source().to_string())
          .unwrap_or(String::from("this is an empty asset"));
        let tag = Path::new(filename)
          .extension()
          .map(|x| x.to_string_lossy())
          .unwrap_or(std::borrow::Cow::Borrowed("txt"));
        Some(format!("```{tag} title={filename}\n{content}\n```"))
      } else {
        None
      }
    })
    .sorted()
    .join("\n\n");
  settings.bind(|| {
    assert_snapshot!(snapshot_name.as_str(), content);
  });
  let stats = compiler.compilation.get_stats();
  let warnings = stats.get_warnings();
  let errors = stats.get_errors();
  if !warnings.is_empty() && errors.is_empty() {
    println!(
      "Warning to compile in fixture {:?}, warnings: {:?}",
      fixture_path,
      stats
        .emit_diagnostics_string(true)
        .expect("failed to emit diagnostics to string")
    )
  }
  if !errors.is_empty() {
    panic!(
      "Failed to compile in fixture {:?}, errors: {:?}",
      fixture_path,
      stats
        .emit_diagnostics_string(true)
        .expect("failed to emit diagnostics to string")
    );
  }
  compiler
}

fn read_dir_reverse(path: &PathBuf) -> Vec<String> {
  let mut result = vec![];
  if let Ok(changed_dir) = std::fs::read_dir(path) {
    changed_dir.for_each(|entry| {
      let entry = entry.expect("TODO:");
      if entry.path().is_file() {
        result.push(entry.path().as_os_str().to_string_lossy().to_string());
      }
      if entry.path().is_dir() {
        result.extend(read_dir_reverse(&entry.path()))
      }
    })
  }
  result
}

enum FsOptionEnum {
  Change(Vec<u8>),
  Create,
  Remove(Vec<u8>),
}

#[tokio::main]
pub async fn test_rebuild_fixture(
  fixture_path: &Path,
  cb: Option<Box<dyn FnOnce(Compiler<AsyncNativeFileSystem>)>>,
) {
  enable_tracing_by_env(&std::env::var("TRACE").ok().unwrap_or_default(), "stdout");

  let (options, plugins) = apply_from_fixture(fixture_path);
  // clean output
  if options.output.path.exists() {
    std::fs::remove_dir_all(&options.output.path).expect("should remove output");
  }
  let mut compiler = Compiler::new(options, plugins, AsyncNativeFileSystem);
  compiler
    .build()
    .await
    .unwrap_or_else(|e| panic!("failed to compile in fixture {fixture_path:?}, {e:#?}"));

  let mut files_map: HashMap<String, FsOptionEnum> = HashMap::new();
  let changed_files: HashSet<String> =
    HashSet::from_iter(read_dir_reverse(&fixture_path.join("changed")).into_iter());
  let removed_files: HashSet<String> =
    HashSet::from_iter(read_dir_reverse(&fixture_path.join("removed")).into_iter());
  let created_files: HashSet<String> =
    HashSet::from_iter(read_dir_reverse(&fixture_path.join("created")).into_iter());

  let mut created_files: HashSet<String> = created_files
    .iter()
    .map(|file_path| {
      let target_path = file_path.replacen("created/", "", 1);
      files_map.insert(target_path.clone(), FsOptionEnum::Create);
      let created_context = std::fs::read(file_path.clone());
      std::fs::write(
        target_path.clone(),
        created_context.expect("changed file do not read"),
      )
      .expect("TODO: panic message");
      target_path
    })
    .collect();
  let changed_files: HashSet<String> = changed_files
    .iter()
    .map(|file_path| {
      let target_path = file_path.replacen("changed/", "", 1);
      let target_context = std::fs::read(target_path.clone());
      files_map.insert(
        target_path.clone(),
        FsOptionEnum::Change(target_context.expect("change file not found")),
      );
      let new_content = std::fs::read(file_path);
      std::fs::write(
        target_path.clone(),
        new_content.expect("changed file do not read"),
      )
      .expect("TODO: panic message");
      target_path
    })
    .collect();
  let removed_files: HashSet<String> = removed_files
    .iter()
    .map(|file_path| {
      let target_path = file_path.replacen("removed/", "", 1);
      let target_context = std::fs::read(target_path.clone());
      files_map.insert(
        target_path.clone(),
        FsOptionEnum::Remove(target_context.expect("change file not found")),
      );
      target_path
    })
    .collect();
  created_files.extend(changed_files.into_iter());
  let changed_files = created_files;
  compiler
    .rebuild(changed_files, removed_files)
    .await
    .unwrap_or_else(|e| panic!("failed to rebuild in fixture {fixture_path:?}, {e:#?}"));

  files_map
    .iter()
    .for_each(|(old_path, old_content)| match old_content {
      FsOptionEnum::Create => {
        std::fs::remove_file(old_path).expect("TODO: panic message");
      }
      FsOptionEnum::Change(old_content) | FsOptionEnum::Remove(old_content) => {
        std::fs::write(old_path.clone(), old_content).expect("TODO: panic message");
      }
    });
  let stats = compiler.compilation.get_stats();
  let output_name = make_relative_from(&compiler.options.output.path, fixture_path);
  let rst = RstBuilder::default()
    .fixture(PathBuf::from(fixture_path))
    .actual(output_name)
    .build()
    .expect("TODO:");

  let errors = stats.get_errors();
  if !errors.is_empty() {
    panic!(
      "Failed to compile in fixture {:?}, errors: {:?}",
      fixture_path,
      stats
        .emit_diagnostics_string(true)
        .expect("failed to emit diagnostics to string")
    );
  }
  rst.assert();

  if let Some(cb) = cb {
    cb(compiler);
  }
}
