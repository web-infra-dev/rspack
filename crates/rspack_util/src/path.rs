use std::path::{Component, Components, Path, PathBuf};

fn normalize(parts: Components, allow_above_root: bool) -> Vec<Component> {
  let mut res = vec![];
  for p in parts {
    match &p {
      Component::CurDir => (),
      Component::ParentDir => {
        if !matches!(
          res.last(),
          Some(Component::ParentDir) | Some(Component::RootDir)
        ) {
          res.pop();
        } else if allow_above_root {
          res.push(p);
        }
      }
      _ => res.push(p),
    }
  }
  res
}

pub fn relative(from: &Path, to: &Path) -> PathBuf {
  if from == to {
    return PathBuf::new();
  }

  let is_from_absolute = matches!(from.components().next(), Some(Component::RootDir));
  let is_to_absolute = matches!(to.components().next(), Some(Component::RootDir));

  // At this point the path should be resolved to a full absolute path, but
  // handle relative paths to be safe

  // Normalize the path
  let from = normalize(from.components(), !is_from_absolute);
  let to = normalize(to.components(), !is_to_absolute);

  let mut from_iter = from.iter();
  let mut to_iter = to.iter();
  let mut common_parts = 0;

  let from_remain = loop {
    match (from_iter.next(), to_iter.next()) {
      (Some(from_part), Some(to_part)) if from_part == to_part => {
        common_parts += 1;
      }
      (None, _) => break 0,
      _ => break from_iter.count() + 1,
    }
  };

  let mut result = PathBuf::new();
  for _ in 0..from_remain {
    result.push("..");
  }

  let to_iter = to.into_iter().skip(common_parts);
  for part in to_iter {
    result.push(part.as_os_str());
  }

  result
}

#[cfg(test)]
mod test {
  use std::path::Path;

  use super::*;

  #[cfg(not(target_os = "windows"))]
  #[test]
  fn test_posix() {
    let test_cases = vec![
      ("/var/lib", "/var", ".."),
      ("/var/lib", "/bin", "../../bin"),
      ("/var/lib", "/var/lib", ""),
      ("/var/lib", "/var/apache", "../apache"),
      ("/var/", "/var/lib", "lib"),
      ("/", "/var/lib", "var/lib"),
      (
        "/foo/test",
        "/foo/test/bar/package.json",
        "bar/package.json",
      ),
      ("/Users/a/web/b/test/mails", "/Users/a/web/b", "../.."),
      ("/foo/bar/baz-quux", "/foo/bar/baz", "../baz"),
      ("/foo/bar/baz", "/foo/bar/baz-quux", "../baz-quux"),
      ("/baz-quux", "/baz", "../baz"),
      ("/baz", "/baz-quux", "../baz-quux"),
      ("/page1/page2/foo", "/", "../../.."),
      // Fix https://github.com/web-infra-dev/rspack/issues/8219
      ("/", "/../maps/main.js.map", "maps/main.js.map"),
    ];

    for (from, to, expected) in test_cases {
      let actual = relative(Path::new(from), Path::new(to))
        .to_string_lossy()
        .to_string();
      assert_eq!(actual, expected.to_string());
    }
  }

  #[cfg(target_os = "windows")]
  #[test]
  fn test_win32() {
    let test_cases = vec![
      ("c:/blah\\blah", "d:/games", "d:\\games"),
      ("c:/aaaa/bbbb", "c:/aaaa", ".."),
      ("c:/aaaa/bbbb", "c:/cccc", "..\\..\\cccc"),
      ("c:/aaaa/bbbb", "c:/aaaa/bbbb", ""),
      ("c:/aaaa/bbbb", "c:/aaaa/cccc", "..\\cccc"),
      ("c:/aaaa/", "c:/aaaa/cccc", "cccc"),
      ("c:/", "c:\\aaaa\\bbbb", "aaaa\\bbbb"),
      ("c:/aaaa/bbbb", "d:\\", "d:\\"),
      ("c:/AaAa/bbbb", "c:/aaaa/bbbb", ""),
      ("c:/aaaaa/", "c:/aaaa/cccc", "..\\aaaa\\cccc"),
      ("C:\\foo\\bar\\baz\\quux", "C:\\", "..\\..\\..\\.."),
      (
        "C:\\foo\\test",
        "C:\\foo\\test\\bar\\package.json",
        "bar\\package.json",
      ),
      ("C:\\foo\\bar\\baz-quux", "C:\\foo\\bar\\baz", "..\\baz"),
      (
        "C:\\foo\\bar\\baz",
        "C:\\foo\\bar\\baz-quux",
        "..\\baz-quux",
      ),
      ("\\\\foo\\bar", "\\\\foo\\bar\\baz", "baz"),
      ("\\\\foo\\bar\\baz", "\\\\foo\\bar", ".."),
      ("\\\\foo\\bar\\baz-quux", "\\\\foo\\bar\\baz", "..\\baz"),
      (
        "\\\\foo\\bar\\baz",
        "\\\\foo\\bar\\baz-quux",
        "..\\baz-quux",
      ),
      ("C:\\baz-quux", "C:\\baz", "..\\baz"),
      ("C:\\baz", "C:\\baz-quux", "..\\baz-quux"),
      ("\\\\foo\\baz-quux", "\\\\foo\\baz", "..\\baz"),
      ("\\\\foo\\baz", "\\\\foo\\baz-quux", "..\\baz-quux"),
      ("C:\\baz", "\\\\foo\\bar\\baz", "\\\\foo\\bar\\baz"),
      ("\\\\foo\\bar\\baz", "C:\\baz", "C:\\baz"),
    ];

    for (from, to, expected) in test_cases {
      let actual = relative(Path::new(from), Path::new(to))
        .to_string_lossy()
        .to_string();
      assert_eq!(actual, expected.to_string());
    }
  }
}
