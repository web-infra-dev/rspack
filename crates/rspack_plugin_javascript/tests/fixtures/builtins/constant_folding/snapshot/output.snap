---
source: crates/rspack_testing/src/run_fixture.rs
---
```js title=main.js
(self['webpackChunkwebpack'] = self['webpackChunkwebpack'] || []).push([["main"], {
"./development.js": (function (__unused_webpack_module, exports, __webpack_require__) {
}),
"./index.js": (function (__unused_webpack_module, exports, __webpack_require__) {
__webpack_require__(/* ./development */"./development.js");
}),

},function(__webpack_require__) {
var __webpack_exec__ = function(moduleId) { return __webpack_require__(__webpack_require__.s = moduleId) }
var __webpack_exports__ = (__webpack_exec__("./index.js"));

}
]);
```
